src/backend/utils/adt/acl.c

```diff
@@ -5076,7 +5076,19 @@ roles_is_member_of(Oid roleid, enum RoleRecurseType type,
 			 */
 			if (otherid == admin_of && form->admin_option &&
 				OidIsValid(admin_of) && !OidIsValid(*admin_role))
-				*admin_role = memberid;
+				{
+					*admin_role = memberid;
+
+					/* 
+					 * Need not to iterate through all the members 
+					 * if roleid is bbf_role_admin and admin_role is found.
+					 */
+					if (sql_dialect == SQL_DIALECT_TSQL &&
+						get_bbf_admin_oid_hook  && roleid == (*get_bbf_admin_oid_hook)())
+					{
+						return NULL;
+					}
+				}
```

Comment:
What if we put a hook in `is_admin_of_role()` instead.
If member to be checked is bbf_role_admin then do a direct check for admin membership using syscache, if direct admin membership does not exists fall back to `roles_is_member_of()`

Reply:
If direct membership doesn't exist then it will follow the previous path which again will iterate through all the members even if it will get the required role which we want to avoid.
Also by adding the hook in role_is_member ensures both the cases:
1. If member to be checked is bbf_role_admin then it is doing direct check for admin membership using syscache and then returning the required role.
2. if direct admin membership does not exists then it is iterating through all the membership and once got that required role returning from there itself without further iteration which we do in role_is_member.

Reply:
for babelfish direct member ship will always exists.

Reply:
Also the current changes breaks how `roles_is_member_of` is defined to work.
`roles_is_member_of` should always returns all the members of the given role with given member ship type recursively.

Reply:
`roles_is_member_of ` will continue to work the same except for admin option case when member to be checked is bbf_role_admin. No exactly getting what will be that benefit of doing it in `is_admin_of_role ` because to add hook there we need to additionally handle checking membership through syscache which we are anyway doing it in `roles_is_member_of ` in our case without any additional handling.

Reply:
`roles_is_member_of` is defined to return the both, all the members of the given role & optionally the admin.
This can easily break if some one starts using this function for both fetching the list of roles and admin of a given role.

Reply:
Plus `roles_is_member_of` is a very hot code path called by many other functions, how do we logically explain this change will not break the other functions ?

Reply:
Added hook in `is_admin_of_role ` to check direct membership there itself.